### PR TITLE
Fix propagation bugs

### PR DESCRIPTION
**Pull request by paulperegud**
Fix task and offer propagation bugs. Some refactoring.

### CODE REVIEW

File: golemproto/golem_service.py [hunk 1]
```diff
@@ -151,14 +155,23 @@ def on_receive_respond_challenge(self, proto, prefix, signature, ethereum_pub_ke
...
         peers = self.app.services.peermanager.peers
         return peers
 
-    def get_peer_by_node_id(self, peers, target):
+    def get_peer_by_node_id(self, peers, raw_target_id):
         for peer in peers:
-            enc = encode_hex(peer.remote_pubkey)
-            if enc == target:
+            enc = peer.remote_pubkey
+            if enc == raw_target_id:
```
Comment (Radagast-red) [inline golemproto/golem_service.py]: wouldn't it be netter to move 'decode_hex' here and pass not decoded id around?

File: golemproto/golem_task.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    node_address = None
+    announcement_hash = None
+    signature = "signature"
+    allowed_peers = "some peers"
+    offers = []
+
+    def __init__(self, price):
+        self.utc_time = datetime.datetime.isoformat(datetime.datetime.utcnow())
+        self.price = price
+        self.announcement_hash = 12312312
```
Comment (Radagast-red) [inline golemproto/golem_task.py]: ??

File: golemproto/golem_task.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+class Announcement(object):
+    utc_time = None
+    price = None
+    requestor_id = None
+    node_id = None
+    node_address = None
+    announcement_hash = None
+    signature = "signature"
+    allowed_peers = "some peers"
```
Comment (Radagast-red) [inline golemproto/golem_task.py]: why init this to random tekst?

File: golemproto/golem_task.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        t.signature = signature
+        t.allowed_peers = allowed_peers
+        return t
+
+    def pick_offers(self):
+        from math import log, trunc
+        n = len(self.offers)
+        if n < 2:
+            return False, []
+        n = 2 << trunc(log(n, 2))
```
Comment (Radagast-red) [inline golemproto/golem_task.py]: What is purpose of this?

File: golemproto/golem_service.py [hunk 5]
```diff
@@ -172,55 +185,72 @@ def evaluate_ann(self, ann):
...
+
+    def send_acks(self, ann, best):
+        for offer in best:
+            peer = self.mb_connect(offer.node_address, decode_hex(offer.node_id))
+            if peer is None:
+                log.warning("Can't ACK provider - can't reconnect back: {}, {}",
+                          offer.node_address, offer.node_id)
+                continue
+            log.warning("sending ack to {}".format(offer.node_id))
+            peer.protocols[GolemProtocol].send_acceptance(offer.offer_hash, 43,
```
Comment (Radagast-red) [inline golemproto/golem_service.py]: 43?

File: golemproto/golem_task.py (hunk 3 above)
Comment (paulperegud) [inline golemproto/golem_task.py]: Just a placeholder for now

File: golemproto/golem_task.py (hunk 4 above)
Comment (paulperegud) [inline golemproto/golem_task.py]: Just an example of function that chooses providers. "Pick first 2^n < len(self.offers)"

File: golemproto/golem_service.py (hunk 5 above)
Comment (paulperegud) [inline golemproto/golem_service.py]: placeholder number

File: golemproto/golem_service.py (hunk 1 above)
Comment (paulperegud) [inline golemproto/golem_service.py]: Good idea, will fix

File: golemproto/golem_task.py (hunk 2 above)
Comment (paulperegud) [inline golemproto/golem_task.py]: As above - placeholder number

**Pull request merged**